pkg/secure/mqttfactory.go

```diff
@@ -110,3 +106,38 @@ func (factory MqttFactory) configureMQTTClientForAuth(secretData *messaging.Secr
 
 	return nil
 }
+
+func (factory MqttFactory) getValidSecretData() (*messaging.SecretData, error) {
+	f := func() (*messaging.SecretData, error) {
+		//get the secrets from the secret provider and populate the struct
+		secretData, err := messaging.GetSecretData(factory.authMode, factory.secretPath, factory.sp)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get secret data from the secret provider, error: %s", err)
+		}
+		//ensure that the authmode selected has the required secret values
+		err = messaging.ValidateSecretData(factory.authMode, factory.secretPath, secretData)
+		if err != nil {
+			return nil, fmt.Errorf("invalid secret data, error: %s", err)
+		} else {
+			return secretData, nil
+		}
+	}
+
+	secretData, err := f()
+	if err != nil {
+		factory.logger.Error(err.Error())
+	} else {
+		return secretData, nil
+	}
+	for {
+		factory.logger.Info("Waiting for the secret creation API call to seed the proper credentials...")
+		if _, ok := <-factory.secretAddedSignal; ok {
+			secretData, err := f()
+			if err != nil {
+				factory.logger.Error(err.Error())
+			} else {
+				return secretData, nil
+			}
+		}
+	}
+}
```

Comment:
This code should be in the calling function. I.e. it goes beyond get and validate. Then the above code be the body of the func instead of in-line define func.